lib/sparse/src/index/search_context.rs

```diff
@@ -130,24 +139,31 @@ impl<'a> SearchContext<'a> {
     /// b,  21, 34, 60, 200
     /// b,  30, 34, 60, 230
     fn advance(&mut self) -> Option<ScoredPointOffset> {
-        let min_record_id = Self::next_min_id(&self.postings_iterators)?;
+        // Get current min record id from all posting list iterators
+        let current_min_record_id = *self.min_record_ids.peek()?;
         let mut score = 0.0;
 
-        // Iterate second time to advance posting iterators
+        // Iterate to advance matching posting iterators
         for posting_iterator in self.postings_iterators.iter_mut() {
             if let Some(element) = posting_iterator.posting_list_iterator.peek() {
                 // accumulate score for the current record id
-                if element.record_id == min_record_id {
+                if element.record_id == current_min_record_id {
                     score += element.weight * posting_iterator.query_weight;
                     // advance posting list iterator to next element
                     posting_iterator.posting_list_iterator.advance();
+                    // pop min record id
+                    self.min_record_ids.pop();
+                    // look-ahead to find next min record id
+                    if let Some(next_element) = posting_iterator.posting_list_iterator.peek() {
+                        self.min_record_ids.push(next_element.record_id);
```

Comment:
I cannot understand why we add record id here, we have already filled `self.min_record_ids` in `SearchContext::new` with all query indices

Reply:
`min_record_ids` tracks the min record point ids of the posting list, not the query indices.

As we `pop` an element for a the posting list, we need to the new `next` min record id.
Each non empty posting list must have an entry in the heap.

Does it help?

Reply:
Yep